### PR TITLE
Connectivity Helper to check and record device reachability

### PR DESCRIPTION
**Pull request by murrant**
Please give a short description what your pull request is for

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

File: LibreNMS/Data/Source/Fping.php [hunk 1]
```diff
@@ -15,16 +15,17 @@
  * GNU General Public License for more details.
  *
  * You should have received a copy of the GNU General Public License
- * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  *
- * @link       https://www.librenms.org
- *
- * @copyright  2020 Tony Murray
+ * @package    LibreNMS
```
Comment (Jellyfrog) [inline LibreNMS/Data/Source/Fping.php:20]: You should fix your template :)
Remove package, it's some old deprecated stuff and update links to https

Comment (murrant) [reply LibreNMS/Data/Source/Fping.php:20]: Well, shoot, now I have to remember where that is :D

File: LibreNMS/Data/Source/Fping.php [hunk 2]
```diff
@@ -67,28 +68,10 @@ public function ping($host, $count = 3, $interval = 1000, $timeout = 500, $addre
...
-
-        preg_match('#= (\d+)/(\d+)/(\d+)%(, min/avg/max = ([\d.]+)/([\d.]+)/([\d.]+))?$#', $output, $parsed);
-        [, $xmt, $rcv, $loss, , $min, $avg, $max] = array_pad($parsed, 8, 0);
 
-        if ($loss < 0) {
-            $xmt = 1;
-            $rcv = 1;
-            $loss = 100;
-        }
+        $response = FpingResponse::parseOutput($process->getErrorOutput(), $process->getExitCode());
```
Comment (Jellyfrog) [inline LibreNMS/Data/Source/Fping.php:72]: I see the old code does it also, but fping prints this on stderr?

Comment (murrant) [reply LibreNMS/Data/Source/Fping.php:72]: Correct, fun stuff eh?

**Comment by Jellyfrog** [comment]
> Unlike other database systems e.g., MySQL and PostgreSQL, you cannot use the ALTER TABLE statement to add a primary key to an existing table.
> 
> You need to follow these steps to work around the limitation:
> 
> First, set the foreign key constarint check off.
> Next, rename the table to another table name (old_table)
> Then, create a new table (table) with exact structure of the table that you have been renamed.
> After that, copy data from the old_table to the table.
> Finally, turn on the foreign key constraint check on

**Comment by murrant** [comment]
@Jellyfrog naw, I'll just change history :D

**Comment by Jellyfrog** [comment]
> @Jellyfrog naw, I'll just change history :D

Nobody will know.

**Comment by murrant** [comment]
Well, now we have created a new timeline.   We just leave the other timeline to rot :D

File: LibreNMS/Polling/ConnectivityHelper.php [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        }
+
+        return $status;
+    }
+
+    public function isSNMPable(): bool
+    {
+        $response = NetSnmp::device($this->device)->get('SNMPv2-MIB::sysObjectID.0');
+
+        return $response->getExitCode() === 0 || $response->isValid();
```
Comment (Jellyfrog) [inline LibreNMS/Polling/ConnectivityHelper.php:138]: || vs &&,

I'm guessing not all devices responds to sysobjectid and therefor wouldn't be "valid"?
Maybe deserves a comment?
There is snmpping also but it just seems to fetch some other oids

**Review by Jellyfrog — APPROVED** [review]

File: LibreNMS/Polling/ConnectivityHelper.php (hunk 3 above)
Comment (murrant) [reply LibreNMS/Polling/ConnectivityHelper.php:138]: This is correct matches with previous behavior.  I was very careful about this part as I messed with this before and had to revert. :D
Actually, old code sent one snmpget, check the return code was 0 if it was great.
If not, it sent another request and checked the output.

**Pull request merged**